cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1087,15 +1088,28 @@ private boolean isValid(RunRecordDetail existing, ProgramRunStatus nextProgramSt
    *
    * @param limit count at most that many runs, stop if there are more.
    */
-  public int countActiveRuns(@Nullable Integer limit) throws IOException {
-    AtomicInteger count = new AtomicInteger(0);
+  public Map<ProgramRunStatus, Integer> countActiveRuns(@Nullable Integer limit) throws IOException {
+    Map<ProgramRunStatus, AtomicInteger> tempMap = new ConcurrentHashMap<>();
     try (CloseableIterator<RunRecordDetail> iterator = queryProgramRuns(
       Range.singleton(getRunRecordNamespacePrefix(TYPE_RUN_RECORD_ACTIVE, null)),
       key -> !NamespaceId.SYSTEM.getNamespace().equals(key.getString(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD)),
       null, limit != null ? limit : Integer.MAX_VALUE)) {
-      iterator.forEachRemaining(m -> count.getAndIncrement());
+      iterator.forEachRemaining(m -> {
+        // if this is a program within a workflow or a preview run, counting is skipped.
+        boolean isInWorkflow = m.getSystemArgs().containsKey(ProgramOptionConstants.WORKFLOW_NAME);
+        boolean skipProvisioning = Boolean.parseBoolean(
+          m.getSystemArgs().getOrDefault(ProgramOptionConstants.SKIP_PROVISIONING, "false"));
+        if (isInWorkflow || skipProvisioning) {
+          return;
+        }
+        tempMap.putIfAbsent(m.getStatus(), new AtomicInteger(0));
+        tempMap.get(m.getStatus()).incrementAndGet();
```

Comment:
Since everything is local, you can just do a get and put if `tempMap` is `Map<ProgramRunStatus, Integer>`

```
tempMap.put(m.getStatus(), tempMap.getOrDefault(m.getStatus(), 0));
```

Reply:
Undo changes in this class.